### ISSUE TITLE
fix(modal): modal does not properly work in a lazy loaded module

### ISSUE THREAD

**Issue by dzonatan** [valor-software/ngx-bootstrap#2356]
To have working modal in a lazy-loaded module you have to import modal module with `forRoot()` in that lazy module which is not correct as `BsModalService` will get a new instance and will not know anything about other modals of other modules (e.g. nesting won't properly work). If you skip the `forRoot()` you will get:

```
ERROR Error: No component factory found for MyModalInLazyLoadedModule. 
Did you add it to @NgModule.entryComponents?
```

That's happening because [ComponentLoaderFactory](https://github.com/valor-software/ngx-bootstrap/blob/development/src/modal/modal.module.ts#L17) is provided only in root module which does not know anything about lazy-loaded injector and factories.

Proper fix would be to have some kind of [overlay service](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/src/modal/modal.ts#L47)  that is provided with every [ModalModule import](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/src/modal/modal.module.ts#L15).

Reproduction - https://stackblitz.com/edit/ngx-bootstrap-playground-mtwpzg

**Comment by nischi** [comment]
Is there a workaround? I have exactly the same issue

**Edit**
I have a workaround. I will use the directive with some ViewChilds.

**Comment by psined1** [comment]
Running into the same issue. Could you share an example please?

**Comment by dzonatan** [comment]
@psined1 Just import `ModalModule.forRoot()` in your lazy-loaded module.

**Comment by psined1** [comment]
yep, did that, works fine

**Comment by julianonunes** [comment]
Also having the same issue. The workaround pointed by @dzonatan fixes it, but I don't know what are the implications of using forRoot in a child module.

**Comment by Strike117** [comment]
Having same issue, is there an update?

**Comment by vincentjames501** [comment]
I wouldn't reimport `ModalModule.forRoot()` as it will create a new BsModalService and then you'll have multiple of these services which are trying to manage some global UI state. It would work fine if you don't plan on using the service to call something like `bsModalService.hide(1)` from another context (which our app does to force close modals upon a navigation event).

I think the easiest solution would be (while still somewhat kludgy), for the lazy loaded module to provide its own `ComponentLoaderFactory` and extend `ModalOptions` to allow us to give our own `ComponentLoaderFactory` as then it would have the correct underlying `ComponentFactoryResolver` which could resolve lazy `entryComponents` (at least according to https://github.com/angular/angular/issues/14324).

**Comment by xino1010** [comment]
advances?

**Comment by mdgilene** [comment]
Any updates on this?

**Comment by jbeckton** [comment]
Nov 2018, still broken?

**Comment by thetric** [comment]
https://github.com/valor-software/ngx-bootstrap/pull/3361 aims to fix it, but according to @valorkin it is [still broken](https://github.com/valor-software/ngx-bootstrap/pull/3361#issuecomment-399044968). Sadly, the modal roadmap issue https://github.com/valor-software/ngx-bootstrap/issues/3258 does not show progress for about a year, too 😢 

Edit: You can use `TemplateRef`s in lazy loaded modules, but not `Component`s

**Comment by fabianedl777** [comment]
Any updates on this?

**Comment by fabianedl777** [comment]
I needed to perform an action every time I opened a modal in a lazy module to be able to do this. I create a shared module and import the modal module and in the constructor add a listener to the event emiter that the BModalService has. In this way I was not forced to be repeating this in all the modules where I wanted to do this.

I hope you can help them this.

**Comment by lambidu** [comment]
Same problem, after almost two years ;(
Check this [reproduction](https://github.com/lambidu/ngx-bootstrap-test), maybe it will help revive this thread.

As a side note, I've tried to apply workaround that @thetric proposed - yes, it works but only until you need `BsModalRef` service injected into your modal to programmatically manipulate/hide dialog. Another propose was to use `<ng-template />` - this will produce a lot of duplicates if you need to reuse it somewhere.

Disappointed...

**Comment by himanshugarg574** [comment]
Any update on the issue?

**Comment by JoshuaAlzate** [comment]
I believe this is an Angular issue [here](https://github.com/angular/angular/issues/14324). But wow, this issue still not closed.

**Comment by maxandriani** [comment]
This is not entirely an angular issue. The way how ng works is not compatible to your guys implementation. The lazy load module encapsulation will never allow provide lazy load factories on global FactoryResolver.

The way ng team and obter libs fixed this issue is working w/ two services. One provided in Module Scope, dedicated to resolve component reference. And a second service provided in root scope to handle multiple modal instances.

In short, the local service will resolve the component reference w/ local factory resolver, and then inject a global service to deal w/ multiple instances of resolved components/modals.

**Comment by conrado-l** [comment]
¿Any updates on this?

I'm getting multiple modalService instances because of the "ModalModule.forRoot()" workaround, and by doing this I can't use "setDismissReason" and the "onHide" event properly for all the modals.

Thanks for the great library.

**Comment by jeserkin** [comment]
Decided to try and extract modal portion into separate component and pass it in. What do you know, same issue as reported by so many others. I did import `ModalModule.forRoot()` into lazy-loaded module. I did declare modal component in `declarations` section and `entryComponents` section of lazy-loaded module. There was also a suggestion on SO, to re-provide `BsModalService`.

Result: **Nothing hepled**
Still receiving 
```
ERROR Error: No component factory found for XComponent. Did you add it to @NgModule.entryComponents?
```

**Comment by MarkoSulamagi** [comment]
Bringing this issue up again.

**Comment by himanshugarg574** [comment]
And again

**Comment by jeserkin** [comment]
Is this issue present in Angular 9? When I was updating from Angular 8.2 to Angular 9.1 it said following.

![image](https://user-images.githubusercontent.com/183512/86437451-4353a080-bd0d-11ea-9cb9-b0da35d56d4c.png)

**Comment by MarkoSulamagi** [comment]
@jeserkin No, it's actually not an issue with Angular 9 :) I got rid of it after the upgrade.

**Comment by jeserkin** [comment]
So to conclude. Update to Angular 9 and `entryComponent` issue is resolved + https://github.com/valor-software/ngx-bootstrap/issues/2356#issuecomment-337816560

**Comment by himanshugarg574** [comment]
But it still generates multiple instances of modalService(one for each lazy loaded module) which is not correct.

**Comment by jeserkin** [comment]
Doesn't tooltip and popover do the same? They also require TooltipModule.forRoot(), PopoverModule.forRoot() as far as I know.
